hibernate-core/src/main/java/org/hibernate/bytecode/enhance/internal/bytebuddy/EnhancerImpl.java

```diff
@@ -532,8 +532,26 @@ String getDescriptor() {
 			return fieldDescription.getDescriptor();
 		}
 
-		boolean isVisibleTo(TypeDescription typeDescription) {
-			return fieldDescription.isVisibleTo( typeDescription );
+		boolean isVisibleTo(TypeDescription type) {
+			final var declaringType = fieldDescription.getDeclaringType().asErasure();
+			if ( declaringType.isVisibleTo( type ) ) {
+				if ( fieldDescription.isPublic() ) {
+					return true;
```

Comment:
Technically a package-protected declaring type would pass the check, but would likely end up hitting the problem we're trying to fix in case of split packages.

Though in practice I doubt package-protected types would work well with Hibernate ORM, so I suppose we can ignore this.